### DISCUSSION TITLE
Dynamic Querystring Parameters

### DISCUSSION THREAD

**Discussion by JeremyBP** [christianhelle/refitter#417] [Ideas]
Hi,

While I'm working on Apizr integration into Refitter, I'm thinking it could be useful to offer the possibility to choose about the query parameters format.
I mean Refit can handle [dynamic querystring parameters](https://github.com/reactiveui/refit?tab=readme-ov-file#dynamic-querystring-parameters), so maybe we should add a new setting like a "DynamicQuerystringParametersThreshold" int property.
A property of int type because we may want to never use it with a 0 value, or use it but starting X query parameters count (1 would be pointless). Or maybe just a boolean with activation at 2.

Anyway, something like this:
```csharp
namespace GeneratedCode
{
    [System.CodeDom.Compiler.GeneratedCode("Refitter", "1.0.0.0")]
    public partial interface IApiClient
    {
        /// <remarks>Update the details of the specified job.</remarks>
        /// <param name="id">Foo Id</param>
        /// <param name="title">Job title</param>
        /// <param name="description">Job description</param>
        /// <param name="contact">Contact Person</param>
        /// <returns>A <see cref="Task"/> that completes when the request is finished.</returns>
        /// <exception cref="ApiException">Thrown when the request returns a non-success status code.</exception>
        [Post("/job/{Id}")]
        Task UpdateJobDetails([AliasAs("Id")] string id, [Query, AliasAs("Title")] string? title = default, [Query, AliasAs("Description")] string? description = default, [Query, AliasAs("Contact")] string? contact = default, CancellationToken cancellationToken = default);
    }
}
```

Could be written like that:
```csharp
namespace GeneratedCode
{
    public class UpdateJobDetailsQuery
    {
        [AliasAs("Title")] 
	public string? Title {get; set;} = default;
		
	[AliasAs("Description")] 
	public string? Description {get; set;} = default;
		
	[AliasAs("Contact")] 
	public string? Contact {get; set;} = default;
    }

    [System.CodeDom.Compiler.GeneratedCode("Refitter", "1.0.0.0")]
    public partial interface IApiClient
    {
        /// <remarks>Update the details of the specified job.</remarks>
        /// <param name="id">Foo Id</param>
        /// <param name="title">Job title</param>
        /// <param name="description">Job description</param>
        /// <param name="contact">Contact Person</param>
        /// <returns>A <see cref="Task"/> that completes when the request is finished.</returns>
        /// <exception cref="ApiException">Thrown when the request returns a non-success status code.</exception>
        [Post("/job/{Id}")]
        Task UpdateJobDetails([AliasAs("Id")] string id, [Query] UpdateJobDetailsQuery? query = default, CancellationToken cancellationToken = default);
    }
}
```

Just wondering :)

**Comment by christianhelle** [comment]
@JeremyBP I love the idea, and it should be trivial to implement. Unfortunately (not for me), I'm starting my summer holiday in a few days and planning to disconnect in the next three weeks. I have an idea of what needs to be done so I might take a look at this before I start my vacation, but I can't promise anything

If you want to give it a go, I'll make sure to find time to review your changes so we can get them in

**Comment by christianhelle** [comment]
@JeremyBP do you happen to know whether Refit supports using immutable records to wrap around the query string parameters? 

I recently introduced [support for generating immutable records](https://github.com/christianhelle/refitter/pull/409) and it would make sense to respect the immutable record option when generating these query string wrappers

**Reply by JeremyBP** [reply]
Yeah actually I started by writing a record query param but, finally changed my sample code to match the one on the Refit readme.
But yes it would be better to use record instead.
Anyway, I'm progressing on Apizr integration and would probably push a PR this week about it.
I'll take a look at that dynamic query param thing after that, if you did not before your holiday disconnection started :p
Glad to hear it's almost holiday time for you :D
I'll have to wait until august on my side, which is not so far anyway \o/

**Comment by JeremyBP** [comment]
While we're closed to reach our goal on that one too, I was wondering if we should update your other little monster aka [REST API Client Code Generator](https://github.com/christianhelle/apiclientcodegen) to give it those new capabilities about both Apizr and Dynamic Querystring Parameters. You tell me :)

**Reply by christianhelle** [reply]
Once we release your Apizr and dynamic query string parameters feature, the REST API Client Code Generator must update the Refitter.Core package and it should work out-of-the-box when working with .refitter files

![image](https://github.com/user-attachments/assets/d9cb0a32-5ab2-4d56-9781-a2bc1d21c7ce)

We could add some additional Apizr specific options in the extensions settings screen, to improve the Add New REST API Client experience

![image](https://github.com/user-attachments/assets/c37511d9-8936-4d3b-aef3-26383be3750c)

I can do that once I'm back on my regular schedule in around 12 days from now.

You're more than welcome to give it a go @JeremyBP, and I'll make sure that we get to merge whatever pull request you create. There's a bit of a learning curve with extending Visual Studio, but if you've tried it before then it should feel familiar.

**Reply by JeremyBP** [reply]
Well I think it could be quite sexy to get "Use Dynamic Querystring Parameters" here in VS options, but it's your call as it's your baby.

About "Use Apizr", as it may depend on "DependencyInjectionSettings", I don't know if it's suitable or not.
I mean setting a "Use Apizr" here to true but no "DependencyInjectionSettings" will work, but generate the static registration helper, not the extended one, which is ok but lets no choice.
Or may be, into Refitter, as I did with UseApizr that initialize a default ApizrSettings, we could introduce a UseDependencyInjection boolean which iinitialize DependencyInjectionSettings with some default basic values. It would let the user call `--use-dependency-injection` with command line and get the basic stuff about it (probably HttpResilience at some point).
Then, we could offer both `UseDependencyInjection` and `UseApizr` from the VS Options window.
Such basic options are here more to let the user know it exists, as in the real world we probably need to push it further and deeper with the help of a .refitter config file.

Anyway, I'll let you work on that one on your own as I need now to focus on my Apizr monster.
I want to make sure I could release its major stable 6 pending version when I'm back home falling August (+ blog post, social network, etc...).
